documentation/proposals/Proposal - Multi-Backend Input.md

```diff
@@ -0,0 +1,190 @@
+    - Input no longer has a hard bond to Windowing. The integration will remain the same to the end user but will use Source Generators, more on that later. 
+    - Input contexts are no longer interfaces, they are instead classes which contain input backends.
+
+## Reference Implementation
+
+Similar to Windowing 3.0, a reference implementation will be included in the main `Silk.NET.Input` package which uses the same API or family of APIs as Windowing. This will be exposed via the `InputBackend` class. The `InputBackend` class contains a static `Create` method which accepts the native handle for the input context on the particular underlying API i.e. on GLFW this handle represents a `GLFWwindow*`. This may be subject to change.
+
+## Source Generator
+
+Alongside the input package there will be a source generator (shipped in a `Silk.NET.Input.Roslyn` NuGet package, which is referenced by the main input package). This source generator will be very small, but it allows us to create a link between windowing and input at compile time without a hard reference.
+
+If the source generator detects that both `Silk.NET.Input` and `Silk.NET.Windowing` are referenced, code with the following API surface will be injected into the assembly:
+```cs
+namespace Silk.NET.Input
+{
+    internal static class InputSurfaceExtensions // internal to avoid conflicts with other assemblies
+    {
+         public static InputContext CreateInput(this ISurface surface);
+    }
+}
+```
+
+The `CreateInput` method will use the surface, obtain its platform data (i.e. by using something like `IGlfwPlatformData`), and then feed that into `InputBackend.Create`. The returned `InputContext` will be configured to have the returned `IInputBackend` as a backend. This method will also appropriately bind the `ISurface.Update` callback:
+- The delegate passed in the `onActivated` parameter of `InputBackend.Create` will bind the `IInputBackend.Update` method to run in the `ISurface.Update` callback.
+- The delegate passed in the `onDeactivated` parameter of `InputBackend.Create` will unbind the `ISurface.Update` callback.
+- These delegates must be called by the reference implementation in its `IInputBackend.Activate` and `IInputBackend.Deactivate` methods (respectively)
+- The `IInputBackend.Activate` and `IInputBackend.Deactivate` methods must be called by the `InputContext.Add` and `InputContext.Remove` methods (respectively)
+
+The API surface for this is defined later in the Proposed API section.
+
+One instance of an `IInputBackend` can only belong to one input context at a time. If multiple input contexts are in use, multiple backend instances should also be used. An input context is set as active/in use by a context using the `Activate` and `Deactivate` methods. `Activate` should throw an exception if already activated, and `Deactivate` should throw an exception if not already activated. Once deactivated, an input backend can be reactivated.
```

Comment:
> An input context is set as active/in use by a context using the `Activate` and `Deactivate` methods
Do you mean an IInputBackend? InputContext doesn't have activate/deactivate methods

It's very unclear to me what the InputContext.Add and Remove do and how it behaves with the backend's activate/deactivate. What happens if I add two backends to an InputContext and activate both, or deactivate both?

Reply:
So the idea is that the average user doesn't actually use `Activate` and `Deactivate`, and that `InputContext.Add` and `InputContext.Remove` implicitly call this. If you try to add an already-activated input context, you can expect an exception. The idea is that an input backend can only be "activated for" one input context at a given time.

Reply:
This is my main concern with this right now - I don't think this Activate/Deactivate makes a lot of sense.
When enquiring about this in a different place you said
> It is intended that backends can be passed between contexts. Will update docs to indicate my intentions.
Apart from me being too stupid to find where you updated this, I don't think that makes sense. Clearly the input context "owns" the backend for the time between Add and Remove.
So why have this? Why not make Add/Remove generic and pass in a type (I'm assuming only one of each type can be activated at once) and instanciate it inside the InputContext and dipose it when InputContext is disposed or the backend is removed. The Activate/Deactivate semantics seem like an unnecessary complication to me.

Reply:
great. can't have two conversations on one line

Reply:
The main reason for Activate and Deactivate is for external control i.e. you have one callback which, when called, you can unbind unrelated resources which will potentially outlive your backend (such as a `ISurface`)

The same could be achieved by exposing appropriate events on `InputContext`, but to use the `ISurface` example once again, we'd need to bind said event to bind the `Update` event when it's added, and then bind the removal event to both remove the `Update` event binding as well as unbind the callback from within the callback, which seems messy. The ownership of a backend is a backend-level construct, not a context-level construct, and we'd have to make a lot of logical compromises to have it defined at the latter.

> Apart from me being too stupid to find where you updated this

iirc we're talking on the line which was inserted to update it.

Reply:
> The main reason for Activate and Deactivate is for external control i.e. you have one callback which, when called, you can unbind unrelated resources which will potentially outlive your backend (such as a ISurface)

I don't understand what this is trying to say?
No one can ever call Activate/Deactivate when using the backend with InputContext. So what is meant with external control?

Reply:
If there can only be one backend in use at a time for a given input context, why not just make it a property?

Otherwise if you add many backends to a context, how would you change which one is active between them? Wouldn't we also have to change the IKeyboard/IMouse/IGamepad/etc lists when the backend is changed?

Reply:
> If there can only be one backend in use at a time for a given input context, why not just make it a property?

The whole idea of this proposal is that you can have multiple backends on an input context, but just one context for an input backend [object].

> I don't understand what this is trying to say?
No one can ever call Activate/Deactivate when using the backend with InputContext. So what is meant with external control?

We need some way of detecting when the backend is no longer connected to an input context, so that we can remove it from the `ISurface` update loop and stop receiving events if we're doing nothing with them. Note that I never said "external control", but referred to usage of "external resources". `ISurface`, here, is an "external resource" which the input backend will be using because of what we do in the `Activated` callback.

Reply:
Hate to nitpick, but you did say external control, but I understand what you are saying now I think.
"external resources" are the typical case for `IDispose`, so what is the advantage of having `Activate` instead of a constructor and `Decativate` instead of `Dispose`?

Reply:
> The whole idea of this proposal is that you can have multiple backends on an input context, but just one context for an input backend [object].
If the current backend in use were a property, the user could set it to change it and since it's a property the ```InputContext``` can react to that change, plus it would make it implicitly clear you can only use one backend at a time. Currently you can add multiple backends, but how do you choose which one is in use? How would input devices be handled? Do these lists change completely when the backend is changed? If so, it completely defeats the purpose of this refactor.

Reply:
> Currently you can add multiple backends, but how do you choose which one is in use? 

All of them are in use, that's the idea. `InputContext.Keyboards` would be all of the `IInputBackend.Keyboards` of the configured backends concatenated together.

> Hate to nitpick, but you did say external control,

Damn well okay that's my bad, I really shouldn't have!

> "external resources" are the typical case for IDispose

Because the lifetimes of the resources in question are not in sync, we're not "cleaning up" after ourselves we're just detaching ourselves of responsibility with particular resources.

Moreover, the input backend actually has no idea what resources it might be bonded do, remember the whole `ISurface` tie is done by a dedicated Windowing-Input integration. The whole point of `Activated` and `Deactivated` (the events, not the methods) is so that _user code_ (or something that represents/helps user code like `surface.CreateInput`) can detect when a backend is in action, so that they can start telling the backend to do stuff (i.e. `Update` its input data)

In something that isn't the Windowing-Input integration, the user for example could listen to `Activated` and, when activated, start a thread for updating the backend. We don't want that thread to be active needlessly, so `Deactivated` tells it to shutdown. The cornerstone of this proposal is that you can morph an input context to contain multiple backends as and when you see fit.

<details>
<summary>A further explainer I wrote before you said you understand what I'm saying, but figured I'd keep it in anyway</summary>

Further explainer: `Activate` is `InputContext`'s way of saying "alright backend, boss is coming look busy" and telling the `IInputBackend` to get its act together and start doing meaningful things (like updating input data). `Activate` (and by extension `Activated`) is a mechanism for telling the backend that we care about it now, which is why it's only at this point that we bond the context to the `ISurface`. `ISurface` will, in all likelihood, outlive the input context (even if by just a couple of nanoseconds in most cases) so we need to be able to detect when `InputContext` no longer cares about a particular backend.

</details>

Reply:
> All of them are in use, that's the idea. ```InputContext.Keyboards``` would be all of the ```IInputBackend.Keyboards``` of the configured backends concatenated together.

This makes no sense to me. What's the point of being able to add/remove backends at will if you'd also need to change all the input device instances you're using? How would you even know which IKeyboard instance to use when you have multiple backends in one context?

Reply:
Multiple keyboards was probably a really poor example here, but the main motivator for this proposal is you'll have different backends for different kinds of input devices. For instance, I plan on adding (at some point in 3.X) a wrapper around OpenXR's input mechanisms. This means that we'd have one input backend to receive traditional keyboard and mice inputs, and then one input backend for VR hands, both of which will be accessible from one object (i.e. in the future we'll have `InputContext.Hands` as well as `InputContext.Keyboards`)

> change all the input device instances you're using?

Given this proposal knows about the original Silk.NET 1.0 input proposal, the same requirement applies:

```
In implementations, once an IJoystick, IGamepad, or other IInputDevice object has been created, it must not be destroyed until its IInputContext is disposed. This is to preserve event bindings.
```

Of course, here `IInputContext` is substituted with `IInputBackend`. 

If the user chooses to remove a backend they've bound events to, that's on them. It'll deactivate, which means all of their event bindings are useless but they knowingly removed the backend. Silk.NET itself likely won't ever remove a backend, only add backends, as I don't see any reason why we'd need to yank one out.

Reply:
@HurricanKai After giving it some thought, I appreciate the merit in using disposables. I'll change the proposal to use `IDisposable`, scrap the whole passing around thing, scrap `Activate` and `Activated` and assume that you'll just need to do any preinit before you add the backend, and add a `Disposing` event. Not the cleanest, but not the dirtiest either.

Reply:
Wait, so one input context can have multiple backends active at the same time? Now this makes more sense.

I'll still bring this up again:

> An input context is set as active/in use by a context using the `Activate` and `Deactivate` methods.

I think you meant to write "An input backend is set..."

Reply:
Yes, that was meant to be backend not context. Sorry.

However, any reference to `Activate` and `Deactivate` has now been removed. Please read the revised lifetime rules introduced in the last two commits.